### PR TITLE
feature: Add more discriminator error messages and return specific er…

### PR DESCRIPTION
**Pull request by rtfpessoa**
…ror when possible

### CODE REVIEW

File: openapi3/schema.go [hunk 1]
```diff
@@ -841,33 +842,53 @@ func (schema *Schema) visitSetOperations(settings *schemaValidationSettings, val
...
+		if schema.Discriminator != nil {
+			pn := schema.Discriminator.PropertyName
+			if valuemap, okcheck := value.(map[string]interface{}); okcheck {
+				discriminatorVal, okcheck := valuemap[pn]
+				if !okcheck {
+					return errors.New("input does not contain the discriminator property")
+				}
+
+				if discriminatorRef, okcheck = schema.Discriminator.Mapping[discriminatorVal.(string)]; len(schema.Discriminator.Mapping) > 0 && !okcheck {
+					return errors.New("input does not a valid discriminator value")
```
Comment (fenollp) [inline openapi3/schema.go:855]: You accidentally a word.

**Comment by fenollp** [comment]
LGTM.

Do please add a test that fits your needs (if you're looking for specific pieces of information out of a schema validation failure, make sure that your test looks for a set of specific strings). This way it's easier to port code and error messages over when this lib switches to a different schema handling lib (for openapi >= 3.1 anyway).

**Comment by rtfpessoa** [comment]
I think regarding changes it should be fine now. I will now work on the tests.

**Comment by rtfpessoa** [comment]
@fenollp I think the tests cover pretty much what I wanted to see. Let me know if you have suggestions on how to clean anything or reorganize.

**Comment by rtfpessoa** [comment]
Hi @fenollp, I am trying to understand from my side if we will need to use one internal version temporarily,
do you have any idea about when you will get time to merge this PR?

**Comment by fenollp** [comment]
Sure! So this seems pretty harmless to me.
The only thing I'm curious about is why the "full details" on that existing test ended up disappearing and is there a way to bring them back. I'm guessing this has to do with failfast?

**Comment by fenollp** [comment]
If this is time sensitive you can always use https://golang.org/ref/mod#go-mod-file-replace

**Comment by rtfpessoa** [comment]
> The only thing I'm curious about is why the "full details" on that existing test ended up disappearing and is there a way to bring them back. I'm guessing this has to do with failfast?

Yup, when fail fast returns it will return a different type of error that will give the extra context.
When you use `SchemaError` it will directly delegate the message to the underlying error.

File: openapi3/schema.go [hunk 2]
```diff
@@ -878,7 +895,10 @@ func (schema *Schema) visitSetOperations(settings *schemaValidationSettings, val
 			}
 			if ok > 1 {
 				e.Origin = ErrOneOfConflict
+			} else if len(validationErrors) == 1 {
+				e.Origin = validationErrors[0]
 			}
```
[the remark this answers is not in the document]
Comment (fenollp) [reply openapi3/schema.go:900]: Well the loss of context isn't that huge, also this should all be replaced by a dedicated lib anyway so let's get on with this.

**Pull request merged**